python/ray/data/_internal/logical/interfaces.py

```diff
@@ -0,0 +1,123 @@
+from typing import List
+
+from ray.data._internal.execution.interfaces import Operator, PhysicalOperator
+from ray.data._internal.execution.legacy_compat import (
+    _blocks_to_input_buffer,
+    _stage_to_operator,
+)
+from ray.data._internal.plan import Stage
+
+
+class LogicalOperator(Operator):
+    """Abstract class for logical operators.
+
+    A logical operator describes transformation, and later is converted into
+    physical operator.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["LogicalOperator"]):
+        super().__init__(name, input_dependencies)
+        for x in input_dependencies:
+            assert isinstance(x, LogicalOperator), x
+
+    def set_input_dependencies(self, input_dependencies: List["LogicalOperator"]):
+        self._input_dependencies = input_dependencies
+
+    def set_legacy_stage(self, stage: Stage):
+        self._stage = stage
+
+
+class LogicalPlan:
+    """The plan with a DAG of logical operators."""
+
+    def __init__(self, dag: LogicalOperator):
+        self._dag = dag
+
+    def add(self, operator: LogicalOperator) -> "LogicalPlan":
+        """Add the operator into DAG of logical operators."""
+        assert self._dag is not None
+        operator.set_input_dependencies([self._dag])
+        self._dag = operator
+        return self
+
+    def get_execution_dag(self) -> PhysicalOperator:
+        """Get the DAG of physical operators to execute.
+
+        This process has 3 steps:
+        (1).logical optimization: optimize logical plan.
+        (2).planning: convert logical to physical plan.
+        (3).physical optimization: optimize physical plan.
+        """
+        logical_optimizer = LogicalOptimizer()
+        optimized_plan = logical_optimizer.optimize(self)
+        physical_plan = logical_optimizer.plan(optimized_plan)
+        final_plan = PhysicalOptimizer().optimize(physical_plan)
+        return final_plan._dag
+
+
+class PhysicalPlan:
+    """The plan with a DAG of physical operators."""
+
+    def __init__(self, dag: PhysicalOperator):
+        self._dag = dag
+
+
+class LogicalOptimizer:
```

Comment:
Make this an interface and move its implementation into a separate file?

Reply:
Added an `Optimizer` interface and moved implementation to `optimizers.py` file.